# **LABORATORIO 9: REDES NEURONALES ARTIFICIALES**

## **PARTE A: Red Neuronal de Clasificación**

### **Preparación de los datos**

In [1]:
# Importación de librerías
import pandas as pd # Trabajar con datos estructurados
import numpy as np
from IPython.display import display, HTML # Visualización de objetos
import matplotlib.pyplot as plt # Visualización de gráficos
import seaborn as sns # Visualización de gráficos
import warnings
warnings.filterwarnings("ignore", category=FutureWarning)

In [2]:
# Lectura de la data
data = pd.read_csv('cancer.csv')
data.head()

,Unnamed: 0,ClumpThick,UnifCellSize,UnifCellShape,MargAdh,SingEpiCellSize,BareNuclei,BlandChromatin,NormalNucleoli,Mitoses,Class
0,0,5,1,1,1,2,1,3,1,1,0
1,1,5,4,4,5,7,10,3,2,1,0
2,2,3,1,1,1,2,2,3,1,1,0
3,3,6,8,8,1,3,4,3,7,1,0
4,4,4,1,1,3,2,1,3,1,1,0


In [3]:
# Eliminacion de primera columna
col = data.columns[0]
data = data.drop(col, axis=1)
data.head()

,ClumpThick,UnifCellSize,UnifCellShape,MargAdh,SingEpiCellSize,BareNuclei,BlandChromatin,NormalNucleoli,Mitoses,Class
0,5,1,1,1,2,1,3,1,1,0
1,5,4,4,5,7,10,3,2,1,0
2,3,1,1,1,2,2,3,1,1,0
3,6,8,8,1,3,4,3,7,1,0
4,4,1,1,3,2,1,3,1,1,0


In [4]:
# Se revisan los valores faltantes
data.isnull().sum()

ClumpThick         0
UnifCellSize       0
UnifCellShape      0
MargAdh            0
SingEpiCellSize    0
BareNuclei         0
BlandChromatin     0
NormalNucleoli     0
Mitoses            0
Class              0
dtype: int64

In [5]:
# Se corrobora el tipo de datos
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 699 entries, 0 to 698
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype
---  ------           --------------  -----
 0   ClumpThick       699 non-null    int64
 1   UnifCellSize     699 non-null    int64
 2   UnifCellShape    699 non-null    int64
 3   MargAdh          699 non-null    int64
 4   SingEpiCellSize  699 non-null    int64
 5   BareNuclei       699 non-null    int64
 6   BlandChromatin   699 non-null    int64
 7   NormalNucleoli   699 non-null    int64
 8   Mitoses          699 non-null    int64
 9   Class            699 non-null    int64
dtypes: int64(10)
memory usage: 54.7 KB


In [6]:
# Se revisan los valores únicos de la variable objetivo 'Class'
data['Class'].unique()

array([0, 1], dtype=int64)

In [7]:
# Se convierte la variable objetivo ('Class') a categórica
data['Class'] = data['Class'].astype('category')

In [8]:
# Se separan las características (X) de la variable objetivo (y)
X = data.select_dtypes(include='number')
y = data['Class']

In [9]:
# Se escalan las características
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

In [10]:
# Se dividen los datos en conjuntos de entrenamiento y prueba
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=42)

### **Modelo de red neuronal**

In [11]:
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score

# Ajuste de hiperparámetros
# Se van a evaluar las siguientes arquitecturas de capas ocultas en la red neuronal:
# 1) 1 capa oculta de 5 neuronas
# 2) 1 capa oculta de 10 neuronas
# 3) 2 capas ocultas de 5 neuronas cada una
# 4) primera capa oculta de 10 neuronas y segunda oculta de 5 neuronas
# 5) primera capa oculta de 5 neuronas y segunda oculta de 10 neuronas
# 6) 2 capas ocultas de 10 neuronas cada una

# Se evaluan 3 cantidades de iteraciones: 200, 300 y 400

# Se prueban 2 funciones de activación:
# 1) Relu: f(x)=max(0,x)
# 2) Tangente hiperbólica: f(x)=tanh(x)

best_accuracy = 0.0 
best_model = None 
hidden_layer_sizes_options = [(5,), (10,), (5, 5), (10, 5), (5, 10), (10, 10)]
max_iter_options = [200, 300, 400]
activation_options = ['relu', 'tanh']
solver_options = ['adam', 'lbfgs'] # Se manejan dos tipos de solver para la actualización de los pesos sinápticos y bias en cada iteración

print("Comenzando la búsqueda de los mejores hiperparámetros...")

for hidden_layers in hidden_layer_sizes_options:
    for max_iter in max_iter_options:
        for activation in activation_options:
            for solver in solver_options:
                mlp = MLPClassifier(hidden_layer_sizes=hidden_layers, max_iter=max_iter,
                                    activation=activation, solver=solver, random_state=42)
                mlp.fit(X_train, y_train)
                y_pred = mlp.predict(X_test)
                accuracy = accuracy_score(y_test, y_pred)
                print(f"Capa(s) oculta(s): {hidden_layers}, Iteraciones: {max_iter}, Activación: {activation}, Solver: {solver}, Precisión: {accuracy:.4f}")
                if accuracy > best_accuracy:
                    best_accuracy = accuracy
                    best_model = mlp
                    best_params = {'hidden_layers': hidden_layers, 'max_iter': max_iter,
                                   'activation': activation, 'solver': solver}

# 5. Evaluar el mejor modelo
print("\nMejor modelo encontrado:")
print(f"Parámetros: {best_params}")
print(f"Precisión en el conjunto de prueba del mejor modelo: {best_accuracy:.4f}")

Comenzando la búsqueda de los mejores hiperparámetros...
Capa(s) oculta(s): (5,), Iteraciones: 200, Activación: relu, Solver: adam, Precisión: 0.3257


C:\Users\ACER\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Capa(s) oculta(s): (5,), Iteraciones: 200, Activación: relu, Solver: lbfgs, Precisión: 0.9600


C:\Users\ACER\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Capa(s) oculta(s): (5,), Iteraciones: 200, Activación: tanh, Solver: adam, Precisión: 0.9600
Capa(s) oculta(s): (5,), Iteraciones: 200, Activación: tanh, Solver: lbfgs, Precisión: 0.9429


C:\Users\ACER\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (300) reached and the optimization hasn't converged yet.
  warnings.warn(


Capa(s) oculta(s): (5,), Iteraciones: 300, Activación: relu, Solver: adam, Precisión: 0.3257
Capa(s) oculta(s): (5,), Iteraciones: 300, Activación: relu, Solver: lbfgs, Precisión: 0.9600


C:\Users\ACER\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (300) reached and the optimization hasn't converged yet.
  warnings.warn(


Capa(s) oculta(s): (5,), Iteraciones: 300, Activación: tanh, Solver: adam, Precisión: 0.9600
Capa(s) oculta(s): (5,), Iteraciones: 300, Activación: tanh, Solver: lbfgs, Precisión: 0.9429


C:\Users\ACER\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (400) reached and the optimization hasn't converged yet.
  warnings.warn(


Capa(s) oculta(s): (5,), Iteraciones: 400, Activación: relu, Solver: adam, Precisión: 0.9429
Capa(s) oculta(s): (5,), Iteraciones: 400, Activación: relu, Solver: lbfgs, Precisión: 0.9600
Capa(s) oculta(s): (5,), Iteraciones: 400, Activación: tanh, Solver: adam, Precisión: 0.9657
Capa(s) oculta(s): (5,), Iteraciones: 400, Activación: tanh, Solver: lbfgs, Precisión: 0.9429


C:\Users\ACER\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Capa(s) oculta(s): (10,), Iteraciones: 200, Activación: relu, Solver: adam, Precisión: 0.9657
Capa(s) oculta(s): (10,), Iteraciones: 200, Activación: relu, Solver: lbfgs, Precisión: 0.9371


C:\Users\ACER\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Capa(s) oculta(s): (10,), Iteraciones: 200, Activación: tanh, Solver: adam, Precisión: 0.9600
Capa(s) oculta(s): (10,), Iteraciones: 200, Activación: tanh, Solver: lbfgs, Precisión: 0.9600
Capa(s) oculta(s): (10,), Iteraciones: 300, Activación: relu, Solver: adam, Precisión: 0.9714
Capa(s) oculta(s): (10,), Iteraciones: 300, Activación: relu, Solver: lbfgs, Precisión: 0.9371
Capa(s) oculta(s): (10,), Iteraciones: 300, Activación: tanh, Solver: adam, Precisión: 0.9657
Capa(s) oculta(s): (10,), Iteraciones: 300, Activación: tanh, Solver: lbfgs, Precisión: 0.9600
Capa(s) oculta(s): (10,), Iteraciones: 400, Activación: relu, Solver: adam, Precisión: 0.9714
Capa(s) oculta(s): (10,), Iteraciones: 400, Activación: relu, Solver: lbfgs, Precisión: 0.9371
Capa(s) oculta(s): (10,), Iteraciones: 400, Activación: tanh, Solver: adam, Precisión: 0.9657
Capa(s) oculta(s): (10,), Iteraciones: 400, Activación: tanh, Solver: lbfgs, Precisión: 0.9600


C:\Users\ACER\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Capa(s) oculta(s): (5, 5), Iteraciones: 200, Activación: relu, Solver: adam, Precisión: 0.9714
Capa(s) oculta(s): (5, 5), Iteraciones: 200, Activación: relu, Solver: lbfgs, Precisión: 0.9543


C:\Users\ACER\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Capa(s) oculta(s): (5, 5), Iteraciones: 200, Activación: tanh, Solver: adam, Precisión: 0.9657
Capa(s) oculta(s): (5, 5), Iteraciones: 200, Activación: tanh, Solver: lbfgs, Precisión: 0.9486
Capa(s) oculta(s): (5, 5), Iteraciones: 300, Activación: relu, Solver: adam, Precisión: 0.9714
Capa(s) oculta(s): (5, 5), Iteraciones: 300, Activación: relu, Solver: lbfgs, Precisión: 0.9543


C:\Users\ACER\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (300) reached and the optimization hasn't converged yet.
  warnings.warn(


Capa(s) oculta(s): (5, 5), Iteraciones: 300, Activación: tanh, Solver: adam, Precisión: 0.9714
Capa(s) oculta(s): (5, 5), Iteraciones: 300, Activación: tanh, Solver: lbfgs, Precisión: 0.9486
Capa(s) oculta(s): (5, 5), Iteraciones: 400, Activación: relu, Solver: adam, Precisión: 0.9714
Capa(s) oculta(s): (5, 5), Iteraciones: 400, Activación: relu, Solver: lbfgs, Precisión: 0.9543


C:\Users\ACER\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (400) reached and the optimization hasn't converged yet.
  warnings.warn(


Capa(s) oculta(s): (5, 5), Iteraciones: 400, Activación: tanh, Solver: adam, Precisión: 0.9657
Capa(s) oculta(s): (5, 5), Iteraciones: 400, Activación: tanh, Solver: lbfgs, Precisión: 0.9486


C:\Users\ACER\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Capa(s) oculta(s): (10, 5), Iteraciones: 200, Activación: relu, Solver: adam, Precisión: 0.9657
Capa(s) oculta(s): (10, 5), Iteraciones: 200, Activación: relu, Solver: lbfgs, Precisión: 0.9600


C:\Users\ACER\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Capa(s) oculta(s): (10, 5), Iteraciones: 200, Activación: tanh, Solver: adam, Precisión: 0.9600
Capa(s) oculta(s): (10, 5), Iteraciones: 200, Activación: tanh, Solver: lbfgs, Precisión: 0.9486
Capa(s) oculta(s): (10, 5), Iteraciones: 300, Activación: relu, Solver: adam, Precisión: 0.9657
Capa(s) oculta(s): (10, 5), Iteraciones: 300, Activación: relu, Solver: lbfgs, Precisión: 0.9600


C:\Users\ACER\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (300) reached and the optimization hasn't converged yet.
  warnings.warn(


Capa(s) oculta(s): (10, 5), Iteraciones: 300, Activación: tanh, Solver: adam, Precisión: 0.9657
Capa(s) oculta(s): (10, 5), Iteraciones: 300, Activación: tanh, Solver: lbfgs, Precisión: 0.9486
Capa(s) oculta(s): (10, 5), Iteraciones: 400, Activación: relu, Solver: adam, Precisión: 0.9657
Capa(s) oculta(s): (10, 5), Iteraciones: 400, Activación: relu, Solver: lbfgs, Precisión: 0.9600
Capa(s) oculta(s): (10, 5), Iteraciones: 400, Activación: tanh, Solver: adam, Precisión: 0.9657
Capa(s) oculta(s): (10, 5), Iteraciones: 400, Activación: tanh, Solver: lbfgs, Precisión: 0.9486


C:\Users\ACER\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\ACER\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:541: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
  self.n_iter_ = _check_optimize_result("lbfgs", opt_res, self.max_iter)


Capa(s) oculta(s): (5, 10), Iteraciones: 200, Activación: relu, Solver: adam, Precisión: 0.9657
Capa(s) oculta(s): (5, 10), Iteraciones: 200, Activación: relu, Solver: lbfgs, Precisión: 0.9486


C:\Users\ACER\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Capa(s) oculta(s): (5, 10), Iteraciones: 200, Activación: tanh, Solver: adam, Precisión: 0.9543
Capa(s) oculta(s): (5, 10), Iteraciones: 200, Activación: tanh, Solver: lbfgs, Precisión: 0.9600
Capa(s) oculta(s): (5, 10), Iteraciones: 300, Activación: relu, Solver: adam, Precisión: 0.9657


C:\Users\ACER\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:541: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
  self.n_iter_ = _check_optimize_result("lbfgs", opt_res, self.max_iter)


Capa(s) oculta(s): (5, 10), Iteraciones: 300, Activación: relu, Solver: lbfgs, Precisión: 0.9429
Capa(s) oculta(s): (5, 10), Iteraciones: 300, Activación: tanh, Solver: adam, Precisión: 0.9600
Capa(s) oculta(s): (5, 10), Iteraciones: 300, Activación: tanh, Solver: lbfgs, Precisión: 0.9600
Capa(s) oculta(s): (5, 10), Iteraciones: 400, Activación: relu, Solver: adam, Precisión: 0.9657
Capa(s) oculta(s): (5, 10), Iteraciones: 400, Activación: relu, Solver: lbfgs, Precisión: 0.9371
Capa(s) oculta(s): (5, 10), Iteraciones: 400, Activación: tanh, Solver: adam, Precisión: 0.9600
Capa(s) oculta(s): (5, 10), Iteraciones: 400, Activación: tanh, Solver: lbfgs, Precisión: 0.9600


C:\Users\ACER\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Capa(s) oculta(s): (10, 10), Iteraciones: 200, Activación: relu, Solver: adam, Precisión: 0.9657
Capa(s) oculta(s): (10, 10), Iteraciones: 200, Activación: relu, Solver: lbfgs, Precisión: 0.9486


C:\Users\ACER\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Capa(s) oculta(s): (10, 10), Iteraciones: 200, Activación: tanh, Solver: adam, Precisión: 0.9600
Capa(s) oculta(s): (10, 10), Iteraciones: 200, Activación: tanh, Solver: lbfgs, Precisión: 0.9543
Capa(s) oculta(s): (10, 10), Iteraciones: 300, Activación: relu, Solver: adam, Precisión: 0.9657
Capa(s) oculta(s): (10, 10), Iteraciones: 300, Activación: relu, Solver: lbfgs, Precisión: 0.9486


C:\Users\ACER\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (300) reached and the optimization hasn't converged yet.
  warnings.warn(


Capa(s) oculta(s): (10, 10), Iteraciones: 300, Activación: tanh, Solver: adam, Precisión: 0.9657
Capa(s) oculta(s): (10, 10), Iteraciones: 300, Activación: tanh, Solver: lbfgs, Precisión: 0.9543
Capa(s) oculta(s): (10, 10), Iteraciones: 400, Activación: relu, Solver: adam, Precisión: 0.9657
Capa(s) oculta(s): (10, 10), Iteraciones: 400, Activación: relu, Solver: lbfgs, Precisión: 0.9486
Capa(s) oculta(s): (10, 10), Iteraciones: 400, Activación: tanh, Solver: adam, Precisión: 0.9600
Capa(s) oculta(s): (10, 10), Iteraciones: 400, Activación: tanh, Solver: lbfgs, Precisión: 0.9543

Mejor modelo encontrado:
Parámetros: {'hidden_layers': (10,), 'max_iter': 300, 'activation': 'relu', 'solver': 'adam'}
Precisión en el conjunto de prueba del mejor modelo: 0.9714


El proceso de búsqueda de hiperparámetros encontró que una red neuronal con una capa oculta de 10 neuronas, utilizando la función de activación ReLU y el optimizador Adam, entrenada durante un máximo de 300 iteraciones, logró una precisión de aproximadamente el 97.14% al clasificar muestras no vistas de cáncer de mama. Este desempeño es muy bueno para el tema de predicción de células cancerígenas.